qualtran/bloqs/arithmetic/addition.py

```diff
@@ -266,6 +268,120 @@ def _add_oop_large() -> OutOfPlaceAdder:
 )
 
 
+@frozen
+class SimpleAddConstant(Bloq):
+    r"""Applies U_{k}|x> = |x + k>.
+
+    Applies addition to input register `|x>` given classical integer 'k'.
+
+    This is the simple version of constant addition because it involves simply converting the
+    classical integer into a quantum parameter and using quantum-quantum addition as opposed to
+    designing a bespoke circuit for constant addition based on the classical parameter.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+        k: The classical integer value to be added to x.
+        cvs: A tuple of control variable settings. Each entry specifies whether that
+            control line is a "positive" control (`cv[i]=1`) or a "negative" control `0`.
+        signed: A boolean condition which controls whether the x register holds a value represented
+            in 2's Complement or Unsigned. This affects the ability to add a negative constant.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+
+    References:
+        [Improved quantum circuits for elliptic curve discrete logarithms](https://arxiv.org/abs/2001.09580) Fig 2a
+    """
+
+    bitsize: int
+    k: int
+    cvs: Tuple[int, ...] = field(converter=lambda v: (v,) if isinstance(v, int) else tuple(v))
+    signed: bool
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        if len(self.cvs) > 0:
+            return Signature(
+                [Register('ctrl', bitsize=len(self.cvs)), Register('x', bitsize=self.bitsize)]
+            )
+        else:
+            return Signature([Register('x', bitsize=self.bitsize)])
+
+    def apply(self, *args) -> Union[int, Iterable[int]]:
+        target_val = args[-1]
+        new_target_val = target_val + self.k
+
+        if self.cvs and args[0] != int(''.join(str(x) for x in self.cvs), 2):
+            new_target_val = target_val
+        ret = (args[0], new_target_val) if self.cvs else (new_target_val,)
+        return ret
+
+    def on_classical_vals(self, *args) -> Dict[str, 'ClassicalValT']:
+        return dict(zip([reg.name for reg in self.signature], self.apply(*args)))
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **regs: SoquetT) -> Dict[str, 'SoquetT']:
+
+        # Assign registers to variables and allocate ancilla bits for classical integer k.
+        x = regs['x']
```

Comment:
i'd personally stil put `x` in the signature and let `**regs` capture the optional control soquet